aana/deployments/idefics_2_deployment.py

```diff
@@ -0,0 +1,259 @@
+import asyncio
+from collections.abc import AsyncGenerator
+from queue import Empty
+from threading import Thread
+from typing import Any
+
+import torch
+import transformers
+from pydantic import BaseModel, Field
+from ray import serve
+from transformers import (
+    AutoModelForVision2Seq,
+    AutoProcessor,
+    TextIteratorStreamer,
+)
+
+from aana.core.models.base import merged_options
+from aana.core.models.chat import ChatMessage
+from aana.core.models.custom_config import CustomConfig
+from aana.core.models.image_chat import ImageChatDialog
+from aana.core.models.sampling import SamplingParams
+from aana.core.models.types import Dtype
+from aana.deployments.base_deployment import BaseDeployment, test_cache
+from aana.deployments.base_text_generation_deployment import ChatOutput, LLMOutput
+from aana.exceptions.runtime import InferenceException
+
+
+async def async_streamer_adapter(streamer):
+    """Adapt the TextIteratorStreamer to an async generator."""
+    while True:
+        try:
+            for item in streamer:
+                yield item
+            break
+        except Empty:
+            # wait for the next item
+            await asyncio.sleep(0.01)
+
+
+class Idefics2Config(BaseModel):
+    """The configuration for the Idefics 2 model.
+
+    Attributes:
+        model (str): the model ID on HuggingFace
+        dtype (str): the data type (optional, default: "auto"), one of "auto", "float32", "float16"
+        enable_flash_attention_2 (bool): enable Flash Attention 2 (optional, default: False)
+        model_kwargs (dict): the extra model keyword arguments
+        do_image_splitting (bool): do image splitting (optional, default: False)
+        default_sampling_params (SamplingParams): the default sampling parameters (optional, default: {"temperature": 0, "max_tokens": 512})
+    """
+
+    model: str
+    model_kwargs: CustomConfig = {}
+    enable_flash_attention_2: bool = Field(default=False)
+    do_image_splitting: bool = Field(default=False)
+    dtype: Dtype = Field(default=Dtype.AUTO)
+    default_sampling_params: SamplingParams = SamplingParams(
+        temperature=1.0, max_tokens=256
+    )
+
+
+@serve.deployment
+class Idefics2Deployment(BaseDeployment):
+    """Deployment to serve the Idefics 2 model."""
+
+    async def apply_config(self, config: dict[str, Any]):
+        """Apply the configuration.
+
+        The method is called when the deployment is created or updated.
+
+        It loads the model and processor from HuggingFace.
+
+        The configuration should conform to the Idefics2Config schema.
+        """
+        config_obj = Idefics2Config(**config)
+
+        self.model_id = config_obj.model
+        self.model_kwargs = config_obj.model_kwargs
+        self.dtype = config_obj.dtype
+        self.default_sampling_params = config_obj.default_sampling_params
+        # if self.dtype == Dtype.INT8:
```

Comment:
Dtype.INT8 isn't valid for this model, please delete this code.

Reply:
Removed